float64 q0
float64 q1
float64 q2
float64 q3
float64 w1
float64 w2
float64 w3